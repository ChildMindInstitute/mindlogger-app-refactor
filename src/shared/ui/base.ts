import { Stack } from '@tamagui/core';
import type { StackProps, TextProps } from '@tamagui/core';
export { Image } from '@tamagui/image';
export { YStack, XStack, ZStack } from '@tamagui/stacks';
export { RadioGroup } from '@tamagui/radio-group';

export { Stack as Box, StackProps as BoxProps, TextProps };
