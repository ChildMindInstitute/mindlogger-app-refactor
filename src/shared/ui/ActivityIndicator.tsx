import { ActivityIndicator } from 'react-native';

import { styled, setupReactNative } from '@tamagui/core';

setupReactNative(ActivityIndicator);

export default styled(
  ActivityIndicator,
  {
    name: 'ActivityIndicator',
    testID: 'loader',
  },
  {
    isReactNative: true,
  },
);
