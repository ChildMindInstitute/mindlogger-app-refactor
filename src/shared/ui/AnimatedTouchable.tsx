import { useMemo } from 'react';

import {
  AnimateProp,
  MotiPressable,
  MotiPressableProps,
} from 'moti/interactions';

import { palette } from '../lib/constants/palette';

type Props = MotiPressableProps & {
  idleStyle?: NonNullable<AnimateProp>;
  pressStyle?: NonNullable<AnimateProp>;
  duration?: number;
};

/**
 * Wrapper component for making an element smoothly transition when pressed. Defaults to
 * a simple color change, between the `surface` and `secondary_container` color tokens.
 *
 * Note: All of Tamagui's base components have built-in support for animated pressStyle,
 * however that seemed to break after the React Native 0.79 upgrade. There are no errors,
 * it's just that Tamagui animations seem to have simply stopped running. Tried with the
 * RN Animated support library (@tamagui/animations-react-native) as well as with the
 * Reanimated support library (@tamagui/animations-moti), no luck with either.
 *
 * So using MotiPressable for now.
 */
export function AnimatedTouchable({
  children,
  idleStyle = {
    backgroundColor: palette.surface,
  },
  pressStyle = {
    backgroundColor: palette.neutral99_and_neutral10_012,
  },
  duration = 200,
  ...props
}: Props) {
  return (
    <MotiPressable
      animate={useMemo(
        () =>
          ({ pressed }) => {
            'worklet';
            return pressed ? pressStyle : idleStyle;
          },
        [pressStyle, idleStyle],
      )}
      transition={{ duration }}
      {...props}
    >
      {children}
    </MotiPressable>
  );
}
