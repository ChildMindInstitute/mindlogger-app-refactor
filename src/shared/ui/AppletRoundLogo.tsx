/* eslint-disable react-native/no-inline-styles */
import React, { FC } from 'react';
import { StyleSheet } from 'react-native';

import { Text } from '@tamagui/core';
import { CachedImage } from 'react-native-img-cache';
import Svg, { Circle, Defs, LinearGradient, Stop } from 'react-native-svg';

import { Applet } from '@entities/applet/lib';

import { colors } from '../lib';

import { Box } from '.';

declare module 'react-native-svg' {
  interface LinearGradientProps {
    children: JSX.Element | JSX.Element[];
  }
}

type Props = {
  applet: Applet;
  size?: number;
};

const getImageUrl = (url: string): string => {
  if (url.endsWith('.svg')) {
    // todo - remove if unnecessary
    return url.replace('.svg', '.jpg');
  }
  return url;
};

const AppletRoundLogo: FC<Props> = ({ applet, size = 64 }) => {
  if (applet.image) {
    return (
      <CachedImage
        style={getStyles(size).image}
        source={{ uri: getImageUrl(applet.image) }}
      />
    );
  }

  return (
    <Box w={size} h={size} jc="center" ai="center">
      <Svg height={size} width={size} style={{ position: 'absolute' }}>
        <Defs>
          <LinearGradient id="grad" x1="0" y1="0" x2={size} y2={size}>
            <Stop offset="0" stopColor={colors.blue2} stopOpacity="1" />
            <Stop offset="1" stopColor={colors.lightGreen} stopOpacity="1" />
          </LinearGradient>
        </Defs>

        <Circle cx={size / 2} cy={size / 2} r={size / 2} fill="url(#grad)" />
      </Svg>

      <Text fontSize={size / 2} color="$whiteTsp" fontWeight="700">
        {applet.name[0].toUpperCase()}
      </Text>
    </Box>
  );
};

const getStyles = (size: number) =>
  StyleSheet.create({
    image: {
      width: size,
      height: size,
      resizeMode: 'cover',
      borderRadius: size / 2,
    },
  });

export default AppletRoundLogo;
