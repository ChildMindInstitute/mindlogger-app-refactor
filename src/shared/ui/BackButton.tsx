import { PropsWithChildren } from 'react';
import { TouchableOpacity } from 'react-native';

import { useNavigation } from '@react-navigation/native';

import { Box, BoxProps } from './base';

type RootParamList = ReactNavigation.RootParamList;

type Props<RouteName extends keyof RootParamList> =
  undefined extends RootParamList[RouteName]
    ?
        | { fallbackRoute?: RouteName }
        | {
            fallbackRoute: RouteName;
            fallbackParams: RootParamList[RouteName];
          }
    : { fallbackRoute: RouteName; fallbackParams: RootParamList[RouteName] };

export function BackButton<TRouteName extends keyof RootParamList>(
  props: PropsWithChildren<Props<TRouteName>> & BoxProps,
) {
  // fallbackParams is of any type here if we use object destructuring
  // @ts-ignore
  const { children, fallbackRoute, fallbackParams, ...styledProps } = props;
  const navigation = useNavigation();

  const back = () => {
    if (navigation.canGoBack()) {
      navigation.goBack();
    } else if (fallbackRoute) {
      // @react-navigation's poor type inference
      // @ts-ignore
      navigation.navigate(fallbackRoute, fallbackParams);
    }
  };

  return (
    <Box {...styledProps}>
      <TouchableOpacity onPress={back} hitSlop={40}>
        {children}
      </TouchableOpacity>
    </Box>
  );
}
