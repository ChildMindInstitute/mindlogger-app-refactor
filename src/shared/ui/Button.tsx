import { PropsWithChildren } from 'react';
import { TouchableOpacity, StyleSheet, ColorValue } from 'react-native';

import { colors } from '../lib';

import { ActivityIndicator, BoxProps, Center, Text } from '.';

type Props = PropsWithChildren<{
  onPress: () => void;
  isLoading?: boolean;
  spinnerColor?: ColorValue;
}> &
  BoxProps;

function Button({ onPress, isLoading, spinnerColor, children, ...styledProps }: Props) {
  return (
    <TouchableOpacity onPress={onPress} style={styles.touchable} disabled={isLoading}>
      <Center py={8} bg={colors.blue2} borderRadius="$10" w="100%" {...styledProps}>
        <Text color="$white" fontWeight="bold" fontSize={17} opacity={isLoading ? 0 : 1}>
          {children}
        </Text>

        {isLoading && <ActivityIndicator position="absolute" t={0} b={0} l={0} r={0} color={spinnerColor} />}
      </Center>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  touchable: {
    width: '100%',
  },
});

export default Button;
