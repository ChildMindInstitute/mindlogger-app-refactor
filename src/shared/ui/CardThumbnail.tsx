import React, { FC } from 'react';
import { StyleSheet, AccessibilityProps } from 'react-native';

import { CachedImage } from '@georstat/react-native-image-cache';

import { Box, BoxProps } from './base';
import { Text } from './Text';

type Props = BoxProps & {
  letter?: string;
  imageUri?: string | null;
  size?: number;
};

const getImageUrl = (url: string): string => {
  if (url.endsWith('.svg')) {
    // todo - remove later if unnecessary
    return url.replace('.svg', '.jpg');
  }
  return url;
};

export const CardThumbnail: FC<Props & AccessibilityProps> = ({
  imageUri,
  letter,
  size = 72,
  ...props
}) => {
  if (imageUri || letter) {
    let content: React.ReactNode;

    if (imageUri) {
      const styles = getStyles(size);

      content = (
        <CachedImage style={styles.image} source={getImageUrl(imageUri)} />
      );
    } else if (letter) {
      content = (
        <Box aria-label="round-logo-default" flex={1} jc="center" ai="center">
          <Text fontSize={(size * 32) / 72} color="$on_primary">
            {letter}
          </Text>
        </Box>
      );
    }

    return (
      <Box
        w={size}
        h={size}
        br={8}
        overflow="hidden"
        {...props}
        style={imageUri ? { backgroundColor: 'transparent' } : undefined}
      >
        {content}
      </Box>
    );
  }

  return null;
};

const getStyles = (size: number) =>
  StyleSheet.create({
    image: {
      width: size,
      height: size,
      resizeMode: 'cover',
    },
  });
