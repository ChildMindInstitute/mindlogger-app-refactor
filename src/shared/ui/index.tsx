import CheckBox from '@react-native-community/checkbox';
import { Stack } from '@tamagui/core';
import type { StackProps, TextProps } from '@tamagui/core';

import Center from './Center';
import Input from './Input';
import KeyboardAvoidingView from './KeyboardAvoidingView';
import Link from './Link';
import RowButton from './RowButton';
import ScrollView from './ScrollView';
import Splash from './Splash';
import StatusBar from './StatusBar';
import SubmitButton from './SubmitButton';
import Text from './Text';

export { Image } from '@tamagui/image';
export { YStack, XStack, ZStack } from '@tamagui/stacks';
export { RadioGroup } from '@tamagui/radio-group';
export * from './icons';
export * from './survey';

export { default as MarkdownView } from './MarkdownView';
export { default as RoundLogo } from './RoundLogo';
export { default as RoundTextNotification } from './RoundTextNotification';
export { default as LoadListError } from './LoadListError';
export { default as ImageBackground } from './ImageBackground';
export { default as ActivityIndicator } from './ActivityIndicator';
export { default as NoListItemsYet } from './NoListItemsYet';
export { default as HorizontalCalendar } from './HorizontalCalendar';
export { default as BackButton } from './BackButton';
export { default as Tooltip } from './Tooltip';
export { default as ListSeparator } from './ListSeparator';
export { default as Dropdown } from './Dropdown';
export * from './ViewSlider';
export { default as ProgressBar } from './ProgressBar';
export * from './Stepper';
export { default as Slider } from './Slider';
export { default as Button } from './Button';
export { default as DateTimePicker } from './DateTimePicker';
export { default as ScrollButton } from './ScrollButton';
export { default as ScrollableContent } from './ScrollableContent';
export { default as TouchableOpacity } from './TouchableOpacity';

export {
  StatusBar,
  Center,
  KeyboardAvoidingView,
  Input,
  CheckBox,
  Stack as Box,
  StackProps as BoxProps,
  TextProps,
  ScrollView,
  RowButton,
  SubmitButton,
  Splash,
  Text,
  Link,
};
