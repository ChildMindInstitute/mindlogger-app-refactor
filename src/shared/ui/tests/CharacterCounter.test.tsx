import React from 'react';

import renderer from 'react-test-renderer';

import TamaguiProvider from '@app/app/ui/AppProvider/TamaguiProvider';
import { CharacterCounterText, CharacterCounter } from '@shared/ui';

describe('CharacterCounter Component', () => {
  it('Should render correctly with default props', () => {
    const tree = renderer
      .create(
        <TamaguiProvider>
          <CharacterCounter numberOfCharacters={50} limit={100} />
        </TamaguiProvider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('Should apply the correct font size', () => {
    const tree = renderer.create(
      <TamaguiProvider>
        <CharacterCounter numberOfCharacters={25} limit={50} fontSize={20} />
      </TamaguiProvider>,
    );

    const textElement = tree.root.findByType(CharacterCounterText);
    expect(textElement.props.style).toEqual(
      expect.objectContaining({ color: '#72777F', fontSize: 20 }),
    );
  });

  it('Should apply the primary color when focused', () => {
    const tree = renderer.create(
      <TamaguiProvider>
        <CharacterCounter numberOfCharacters={10} limit={20} focused={true} />
      </TamaguiProvider>,
    );

    const textElement = tree.root.findByType(CharacterCounterText);
    expect(textElement.props.style).toEqual(
      expect.objectContaining({ color: '#0067A0', fontSize: 14 }),
    );
  });

  it('Should apply the grey color when not focused', () => {
    const tree = renderer.create(
      <TamaguiProvider>
        <CharacterCounter numberOfCharacters={10} limit={20} focused={false} />
      </TamaguiProvider>,
    );

    const textElement = tree.root.findByType(CharacterCounterText);
    expect(textElement.props.style).toEqual(
      expect.objectContaining({ color: '#72777F', fontSize: 14 }),
    );
  });

  it('Should display the correct character count', () => {
    const tree = renderer.create(
      <TamaguiProvider>
        <CharacterCounter numberOfCharacters={150} limit={200} />
      </TamaguiProvider>,
    );

    const textElement = tree.root.findByType(CharacterCounterText);

    const characterCountText = Array.isArray(textElement.props.children)
      ? textElement.props.children.join('')
      : textElement.props.children;

    expect(characterCountText).toBe('150/200 character_counter:characters');
  });
});
