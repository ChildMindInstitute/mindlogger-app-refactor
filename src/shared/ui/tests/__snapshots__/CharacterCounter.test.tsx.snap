// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCounter Component Should render correctly with default props 1`] = `
<Text
  focusable={true}
  style={
    [
      {
        "borderBottomWidth": 0,
        "borderLeftWidth": 0,
        "borderRightWidth": 0,
        "borderStyle": "solid",
        "borderTopWidth": 0,
        "color": "#72777F",
        "flex": 1,
        "fontSize": 14,
        "marginBottom": 2,
        "marginLeft": 2,
        "marginRight": 10,
        "marginTop": 2,
        "paddingBottom": 2,
        "paddingLeft": 2,
        "paddingRight": 2,
        "paddingTop": 2,
      },
    ]
  }
>
  50
  /
  100
   
  character_counter:characters
</Text>
`;

exports[`CharacterCounter Component Should renders correctly with default props 1`] = `
<Text
  color="grey4"
  focusable={true}
  fontSize={14}
  style={
    [
      {
        "borderBottomWidth": 0,
        "borderLeftWidth": 0,
        "borderRightWidth": 0,
        "borderStyle": "solid",
        "borderTopWidth": 0,
        "flex": 1,
        "fontWeight": "400",
        "marginBottom": 2,
        "marginLeft": 2,
        "marginRight": 10,
        "marginTop": 2,
        "paddingBottom": 2,
        "paddingLeft": 2,
        "paddingRight": 2,
        "paddingTop": 2,
      },
    ]
  }
>
  50
  /
  100
   
  character_counter:characters
</Text>
`;
